File: mobile/src/types/appRating.ts
export type AppRatingDimensionKey =
  | 'usability_rating'
  | 'helpfulness_rating'
  | 'calendar_rating'
  | 'alerts_rating'
  | 'goals_rating'
  | 'reports_rating'
  | 'records_rating';

export interface AppRatingDto {
  id: number;
  usability_rating: number;
  helpfulness_rating: number;
  calendar_rating: number;
  alerts_rating: number;
  goals_rating: number;
  reports_rating: number;
  records_rating: number;
  suggestions?: string | null;
  created_at: string;
}

export interface AppRatingsSummaryDto {
  total_responses: number;
  averages: {
    usability: number;
    helpfulness: number;
    calendar: number;
    alerts: number;
    goals: number;
    reports: number;
    records: number;
  };
}

export type CreateAppRatingPayload = {
  [key in AppRatingDimensionKey]: number;
} & {
  suggestions?: string;
};

export interface CreateAppRatingResponse {
  id: number;
  message: string;
  created_at: string;
}

File: backend/api_divida_zero/app/models/app_rating.rb
class AppRating < ApplicationRecord
  belongs_to :user

  validates :usability_rating, :helpfulness_rating, :calendar_rating, :alerts_rating,
            :goals_rating, :reports_rating, :records_rating,
            numericality: {
              only_integer: true,
              greater_than_or_equal_to: 1,
              less_than_or_equal_to: 5,
              message: "deve ser uma nota entre 1 e 5."
            }

  validates :suggestions, length: { maximum: 1000, message: "deve ter no máximo %{count} caracteres." }, allow_blank: true

  scope :recent_first, -> { order(created_at: :desc) }

  def serialize
    {
      id: id,
      usability_rating: usability_rating,
      helpfulness_rating: helpfulness_rating,
      calendar_rating: calendar_rating,
      alerts_rating: alerts_rating,
      goals_rating: goals_rating,
      reports_rating: reports_rating,
      records_rating: records_rating,
      suggestions: suggestions,
      created_at: created_at
    }
  end
end
